File: training/train on 1600-1800 problems/C_Travelling_Salesman_and_Special_Numbers.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

int c[1001][1001];
int mod=1e9+7;
int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    string n;
    int k;
    cin>>n;
    cin>>k;
    int l=n.length();
    if (k==0) {
        cout<<1;
    }else if (k==1){
        cout<<l-1;
    }else{

        for (int i=0; i<=1000; i++) {
            c[i][0]=1;
            for (int j=1; j<=i; j++) {
                c[i][j]=(c[i-1][j-1] + c[i-1][j])% mod;
            }
        }

        vector<ll>a (l+1,0);
        for (int i=1;i<l+1;i++) {
        ll res=0;
        int ones=0;
        for (int j=0;j<l;j++) {
            if (n[j]=='1') {
                int need=i-ones;
                if (need>=0 && need<=l-1-j) {
                    res=(res+c[l-1-j][need]);
                }
                ones++;
            }
        }
        if (ones==i) res=(res+1);
        a[i-1]=res;
        }
        vector<int> dp(l+1,0);
        ll res=0;
        for (int i=2;i<l+1;i++){
            dp[i]=1+dp[__builtin_popcount(i)];
            if (dp[i]+1==k){
                res=(ll)(res+a[i-1])%mod;
            }
        }
        cout<<res;
    }
    return 0;
}

File: training/train on 1600-1800 problems/C_Travelling_Salesman_and_Special_Numbers.py
import math

n=input()
k=int(input())
l=len(n)
if k==0:
    print(1)
elif k==1:
    print(l-1)
else:
    c=[[0]*1001 for _ in range(1001)]

    for i in range(1001):
        c[i][0]=1
        for j in range(1,i+1):
            c[i][j]=(c[i-1][j-1]+c[i-1][j])%1000000007

    a=[0]*(l+1)
    for i in range(1,l+1):
        res=0
        ones=0
        for j in range(l):
            if n[j]=="1":
                need=i-ones
                if 0<=need<=l-j-1:
                    res+=c[l-j-1][need]
                ones+=1
        if ones==i:
            res+=1
        a[i-1]=res

    dp=[0]*(l+1)
    res=0
    for i in range(2,l+1):
        dp[i]=1+dp[i.bit_count()]
        if dp[i]+1==k:
            res=(res+a[i-1])%((10**9)+7)

    print(res)

































# import math

# n=int(input(),2)
# k=int(input())
# if k==0:
#     print(1)
# elif k==1:
#     print(n.bit_length()-1)
# else:
#     l=n.bit_length()
#     a=[0]*(l+1)
#     for i in range(1,l+1):
#         res=0
#         ones=0
#         for j in range(l-1,-1,-1):
#             if ((n>>j)&1)==1:
#                 need=i-ones
#                 if 0<=need<=j:
#                     res+=math.comb(j,need)
#                 ones+=1
#         if ones==i:
#             res+=1
#         a[i-1]=res

#     dp=[0]*(l+1)
#     res=0
#     for i in range(2,l+1):
#         dp[i]=1+dp[i.bit_count()]
#         if dp[i]+1==k:
#             res=(res+a[i-1])%((10**9)+7)

#     print(res)
